File: 2025/03/main.go
package main

import (
	"2025/lib"
	"fmt"
	"strconv"
	"strings"
)

var pl = fmt.Println

func printArr(mm []M) {
	for _, x := range mm {
		fmt.Print(x.value)
	}
	fmt.Print("\n")
	fmt.Print("\n")
}

func part1() {
	s := lib.ReadInput("03/example.txt")
	// s := lib.ReadInput("03/input.txt")
	rows := strings.Split(s, "\n")
	maxes := make([]int, len(rows))
	for i, row := range rows {
		maxes[i] = 0
		for a := 0; a < len(row); a++ {
			for b := a + 1; b < len(row); b++ {
				sa := string(row[a])
				sb := string(row[b])
				value, _ := strconv.Atoi(sa + sb)
				if value > maxes[i] {
					maxes[i] = value
				}
			}
		}
	}
	sum := 0
	for i := 0; i < len(maxes)-1; i++ {
		sum += maxes[i]
	}
	// pl(sum)
}

type M struct {
	index int
	value int
}

func toInt64(mm []M) int64 {
	str := ""
	for _, v := range mm {
		str += strconv.Itoa(v.value)
	}
	value, _ := strconv.ParseInt(str, 10, 64)
	return value
}

func filterRows(rows []string) []string {
	arr := make([]string, 0)
	for _, r := range rows {
		if len(r) > 0 {
			arr = append(arr, r)
		}
	}
	return arr
}

func part2() {
	batCount := 12
	// s := lib.ReadInput("03/example.txt")
	s := lib.ReadInput("03/input.txt")
	rows := strings.Split(s, "\n")
	rows = filterRows(rows)

	var sum int64 = 0

	for _, row := range rows {
		// return

		arr := make([]M, 0)
		for len(arr) < batCount {
			m := M{
				index: -1,
				value: -1,
			}
			i := 0
			if len(arr) > 0 {
				i = arr[len(arr)-1].index + 1
			}
			for ; i < len(row)-(batCount-len(arr))+1; i++ {
				v, _ := strconv.Atoi(string(row[i]))
				if v > m.value {
					m = M{
						index: i,
						value: v,
					}
				}
			}
			arr = append(arr, m)
		}
		// pl(toInt64(arr))
		sum += toInt64(arr)
	}
	pl(sum)
}

func main() {
	// part1()
	part2()
}

File: 2025/03/rust/src/main.rs
use std::fs;

fn read_input(file_path: &str) -> String {
    let text = fs::read_to_string(file_path).expect("virhe tiedoston lukussa");
    text
}

fn part1(num_rows: &Vec<Vec<u64>>) {
    let mut biggest_nums: Vec<u64> = Vec::new();

    for row in num_rows {
        let mut max = 0;
        let mut a = 0;
        while a < (row.len() - 1) {
            let mut b = a + 1;
            while b < row.len() {
                let current: u64 = format!("{}{}", row[a], row[b]).parse().expect("virhe pppp");
                // println!("{:?}, {:?} = {:?}, {:?}", a, b, row[a], row[b]);
                if current > max {
                    max = current;
                }
                b += 1;
            }
            a += 1;
        }
        biggest_nums.push(max);
    }
    
    let sum: u64 = biggest_nums.iter().sum();
    // println!("{:?}", biggest_nums);
    println!("{:?}", sum);
}
fn part2(num_rows: &Vec<Vec<u64>>) {
/*
    987654321111111
    811111111111119
    234234234234278
    818181911112111
    
    5 - 3
    23423
*/
    let mut biggest_nums: Vec<u64> = Vec::new();
    let num_size = 12;

    for row in num_rows {
        let mut found: Vec<u64> = Vec::new();
        let mut start_index = 0;
        // find biggest chars in row
        while found.len() < num_size {
            let mut biggest: u64 = 0;
            let mut i = start_index;
            // find biggest char
            while i <= row.len() - (num_size - found.len()) {
                if row[i] > biggest {
                    biggest = row[i];
                    start_index = i + 1;
                }
                i += 1;
            }
            found.push(biggest);
        }
        // println!("found: {:?}", found);
        let mut num_str = String::new();
        for e in found {
            let p = &e.to_string();
            num_str.push_str(p);
        }
        // println!("num_str: {:?}", num_str);
        
        let num: u64 = num_str.parse().expect("virhe löydetyn numberon parsinnassa");
        biggest_nums.push(num);
    }

    
    let sum: u64 = biggest_nums.iter().sum();
    // println!("{:?}", biggest_nums);
    println!("sum: {:?}", sum);
}

fn main() {
    // let text = read_input("../example.txt");
    let text = read_input("../input.txt");
   
    let num_rows: Vec<Vec<u64>> = text.split('\n')
        .filter(|x| x.len() > 0)
        .map(|row| row
            .split("")
            .filter(|x| x.len() > 0)
            .map(|x| x.parse().expect("virhe parsinnassa"))
            .collect()
        ).collect();
    // println!("{:?}", cc);
   
    
    println!("part1 ========================================");
    part1(&num_rows);
    println!("part2 ========================================");
    part2(&num_rows);

}
